eclipse-theia/theia
# electron: fix backend forking

#### What it does

This PR fixes 3 issues:

1. When forking, Electron 4 forgets to set `process.versions.electron` in
the sub-process, making some packages wrongly guess their environment.
This commit manually patches this variable.

2. The generated application code also used to not fork the backend when
working from sources. Running the backend in the Electron main process
was only meant to help with debugging. This commit makes the generated
code fork the backend unless a `--no-cluster` flag is specified.

3. When running a bundled application, the backend was forked. But any
command line parameter passed to the Electron application wasn't passed
down to the backend. This commit makes sure that relevant arguments
are passed to the forked backend process.
\
If you want the generated code to not fork the backend in a sub-process,
you can either pass a `--no-cluster` flag to your application, or set
`THEIA_ELECTRON_NO_BACKEND_FORK=1` environment variable.

Closes: https://github.com/eclipse-theia/theia/pull/7361
Fixes: #7358
Fixes: #4288
Fixes: #4875

#### How to test

Add `grpc` as a dependency, and add `require('grpc')` in `examples/electron/src-gen/backend/server.js`. If you place the require before the patching of the `process.versions.electron` it should not work, but it should work if you put the require after.

note: `grpc` might need to be configured to run with Electron. What I did was:

```sh
cd node_modules/grpc
yarn node-pre-gyp --runtime=electron --target=4.2.12 install
```

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
